engine/src/main/java/net/jqwik/engine/providers/EmailArbitraryProvider.java

```diff
@@ -0,0 +1,23 @@
+package net.jqwik.engine.providers;
+
+import net.jqwik.api.*;
+import net.jqwik.api.constraints.*;
+import net.jqwik.api.providers.*;
+
+import java.util.*;
+
+public class EmailArbitraryProvider implements ArbitraryProvider {
+
+	@Override
+	public boolean canProvideFor(TypeUsage targetType) {
+		Optional<Email> annotation = targetType.findAnnotation(Email.class);
+		return !annotation.equals(Optional.empty());
+	}
```

Comment:
I don't know if this is correct, because I have the following problem:

```
@Property(tries = 10)
void containsAtSign2(@ForAll String email) {
   System.out.println(email);
   //assertThat(email).contains("@");
}
```

Prints now strings like `ጆ衠⋓铨 䟹蘈쭱 禛` or `ꖇ≲`

When I add the Email annotation it prints sometimes correct addresses but sometimes wrong strings. Example:
```
"-EuEa#^A}"@[a04e:6:B:A:9:0:954:4d3c]
쯡᱾ 꺒뮭㜻衔ᴗ 崅ᣆ袽抜 咘周ɠ灀 ꤌꞿ鳘 ා븢 ∟끏豕 푟 腞☉쵊뢣㫾夬  탕娆⑩ﲏ 䚆ᱭ྄멈쭹✺ 쟤 㑸녴茲堡뼥 䄗 逴獷᧾ ⟙ሦ핦▉먲࡞䜘 筝霞㭏뉔꡻ 嶃  梗ȇ귕ƪᇰ
dW_-c&@[D:F:ED:F:C0c:ae::0]
帠ᮩ 較䘐אּ꬗㰓
```

Reply:
The effect you see is that more than one arbitrary provider "declares itself" as suitable. In addition to yours it's `StringArbitraryProvider`. Therefore, jqwik chooses randomly among email and plain string generation.

What you have to do is to override method `int priority()` with a value > 0; I suggest 5.
Moreover, the canProvideFor-method should also check for type String. Similar to this:

```
@Override
public boolean canProvideFor(TypeUsage targetType) {
	return targetType.isAssignableFrom(String.class) && 
				   targetType.findAnnotation(Email.class).isPresent();
}

@Override
public int priority() {
    return 5;
}
```

Reply:
For `@ForAll @Email String email` it works now. But for `@ForAll String email` it prints again strings like
```
芺㳾 릔࠾ ࡊ 䇞
王ᩥ氁ᧂ숪  
釕앾
 宫 
儅䍹攓䄽
 뚄땎ꙺ 
ﺩ툴醨췝㢅敌뗧
෢ 痩닖⹲熌  鴣Ὴ
힕쑸䢲  氁䣜ω㎃ 
 竨聵ጰ洕  䋖䒈籪릲侓◦
```

Reply:
I see no way how `@ForAll String email` could _reliably_ use the email arbitrary. The compilation name of a parameter is only available if you switch it on with a dedicated compiler switch (`-parameters`). That's why you cannot rely on that.

Reply:
I think I didn't explain it clearly enough. The problem is that `@ForAll String string` prints such a string. So the `@ForAll String` breaks and I don't know why.

Reply:
I've now seen version 1.3.8 print the same strings. Maybe that's not a problem at all? I thought it was a problem.

Reply:
An unrestricted string can generate most unicode codepoints. Since there are many more Chinese ones than others those are the most probable to be generated. So this is expected behaviour.

Reply:
Okay, thanks :) Then I had a mistake in my thinking.